nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py

```diff
@@ -106,38 +106,49 @@ def __init__(
 
         special_tokens_dict = {}
-        if self.tokenizer.unk_token is None:
+
+        # # setting special tokens, by default the default model's special tokens will be preserved
+        # # unless passes new values to the special tokens
+        if unk_token is not None:
             special_tokens_dict["unk_token"] = unk_token
-        if self.tokenizer.sep_token is None:
-            if self.tokenizer.eos_token:
-                special_tokens_dict["sep_token"] = self.tokenizer.eos_token
-            else:
-                special_tokens_dict["sep_token"] = sep_token
-        if self.tokenizer.mask_token is None:
+        if mask_token is not None:
             special_tokens_dict["mask_token"] = mask_token
-        if self.tokenizer.bos_token is None:
-            if self.tokenizer.cls_token:
-                special_tokens_dict["bos_token"] = self.tokenizer.cls_token
-            else:
-                special_tokens_dict["bos_token"] = bos_token
-        if self.tokenizer.eos_token is None:
-            if self.tokenizer.sep_token:
-                special_tokens_dict["eos_token"] = self.tokenizer.sep_token
-            else:
-                special_tokens_dict["eos_token"] = eos_token
-        if self.tokenizer.pad_token is None:
+        if pad_token is not None:
             special_tokens_dict["pad_token"] = pad_token
-        if self.tokenizer.cls_token is None:
-            if self.tokenizer.bos_token:
-                special_tokens_dict["cls_token"] = self.tokenizer.bos_token
-            else:
-                special_tokens_dict["cls_token"] = cls_token
 
-        logging.info(f'Adding special tokens to the tokenizer: {special_tokens_dict}')
+        if sep_token is not None:
+            special_tokens_dict["sep_token"] = sep_token
+        elif self.tokenizer.sep_token is None and self.tokenizer.eos_token:
+            special_tokens_dict["sep_token"] = self.tokenizer.eos_token
+        if bos_token is not None:
+            special_tokens_dict["bos_token"] = bos_token
+        elif self.tokenizer.bos_token is None and self.tokenizer.cls_token:
+            special_tokens_dict["bos_token"] = self.tokenizer.cls_token
+        if eos_token is not None:
+            special_tokens_dict["eos_token"] = eos_token
+        elif self.tokenizer.eos_token is None and self.tokenizer.sep_token:
+            special_tokens_dict["eos_token"] = self.tokenizer.sep_token
+        if cls_token is not None:
+            special_tokens_dict["cls_token"] = cls_token
+        elif self.tokenizer.cls_token is None and self.tokenizer.bos_token:
+            special_tokens_dict["cls_token"] = self.tokenizer.bos_token
+
+        new_tokens_in_vocab = []
+        for token in [mask_token, bos_token, eos_token, pad_token, sep_token, cls_token, unk_token]:
+            if token is not None and token not in self.tokenizer.get_vocab():
+                new_tokens_in_vocab.append(token)
```

Comment:
Do we have a plan how to use new_tokens_in_vocab? Shouldn't we add the new tokens to the vocabulary here? Then update the embedding automatically when we see the vocab size does not match embedding size?

Reply:
The new tokens will be added to the vocab since they're included to the special_tokens_dict but we can't resize the model here because the tokenizer doesn't have access to the model. I added the vocab_size() method so that the user can resize the model after the tokenizer is initialized. Also added a section in the tokenizer tutorial about that. I thought that this what we agreed on during the discussion.

Reply:
Yes, but can we add something to the init of the model so that when model finds out that vocab size does is larger than its embedding, expand the embedding automatically or at least give a warning to the user and ask the user to do it by giving a rf to the tutorial?